### DISCUSSION TITLE
Error is triggered for different field types

### DISCUSSION THREAD

**Discussion by nimeshmaharjan1** [react-hook-form/react-hook-form#9559] [Q&A] [answered]
I have a deep nested form and the final field array has a input field which is related with the type recieved from the dropdown before it.
So, I have a URL field type and have a rule to validate the url. When the field type changes to suppose a phone number field the url validation error still shows up.

I would like to appologize as english is not my first language so I am having some difficulty explaining this please take a look at the pictures below 

Any help or suggesstions for anything that I am doing wrong would be very appreciated.

**Firstly, this is the I have selected the URL field type from the dropdown -->**

![urlerror](https://user-images.githubusercontent.com/86953478/206613620-f8678b36-9eaa-4177-991a-e5cc65cd5110.png)

**And have the following rules for the URL input field**

![image](https://user-images.githubusercontent.com/86953478/206614009-8a6b0640-7b77-49e6-92d8-7ad7dcd5c30e.png)

**Changing the field type from URL to phone number**

![phoneerror](https://user-images.githubusercontent.com/86953478/206613656-70abc92e-ab42-4a30-9ae5-e3277f49016f.png)

**And the following rules for the phone  number input field**

![image](https://user-images.githubusercontent.com/86953478/206614119-ac8139ad-a17f-4686-9416-204150bc76f0.png)

**But when I change the field type from URL to phone number I am still getting the URL pattern validation so this url pattern validation keeps getting triggered when I try to submit the form**

![urlphoneerror](https://user-images.githubusercontent.com/86953478/206613663-ce0ddaa5-ad19-4d39-9cdf-44f2b06fdd4e.png)

**Comment by Moshyfawn** [comment, marked as answer]
It's very hard for me to know exactly what's going on in the form, as there's a ton of text and pictures and not enough code _(a minimal repo sandbox would be a lot easier to go through ;))_, but if I understood the behaviour you're after, you want to utilize the `deps` prop to indicate a validation is required on input type value change.

I can't exactly repro your use-case for the lack of implementation details, but here's an example and [a small sandbox repro](https://codesandbox.io/s/vibrant-heyrovsky-3rxoke?file=%2Fsrc%2FApp.tsx%3A964-1589)

```jsx
<select {...register("inputType", { deps: "value" })}>
  {inputTypes.map((inputType) => (
    <option key={inputType} value={inputType}>
      {inputType}
    </option>
  ))}
</select>
{inputType === "tel" && (
  <input
    {...register("value", {
      pattern: { message: "Not a phone", value: telRegex }
    })}
    type="tel"
  />
)}
{inputType === "url" && (
  <input
    {...register("value", {
      pattern: { message: "Not a URL", value: urlRegex }
    })}
    type="url"
  />
)}
```

**Reply by nimeshmaharjan1** [reply]
Firstly thank you so much for looking into this. So  whats happening in my case is that the URL validation is being triggered even when the inputType is "tel".

**Reply by nimeshmaharjan1** [reply]
Thank you so much I just had to pass the shouldUnregister prop for each field

**Comment by nimeshmaharjan1** [comment]
This is basically a part of the code like you have demonstrated

```js
const field = watch(
    `information_collectors[${parentIndex}].question_module.answers[${answersIndex}].modules[${index}].custom_field_object`
  );
  const fieldType = field.format_type;
```

```js
{fieldType === 'URL' && (
        <Controller
          control={control}
          name={toRegister}
          rules={{
            required: 'URL field cannot be empty.',
            pattern: {
              value:
                /(?:https?):\/\/(\w+:?\w*)?(\S+)(:\d+)?(\/|\/([\w#!:.?+=&%!\-/]))?/,
              message: 'This must be a URL address.',
            },
          }}
          render={({
            field: { value, name, ref, onChange },
            fieldState: { error },
          }) => (
            <Form.Control
              style={{ fontSize: 13, height: 33 }}
              placeholder="Enter URL here..."
              className={`${error ? 'is-invalid border-red' : ''}`}
              {...{ value, name, ref, onChange }}
            ></Form.Control>
          )}
        ></Controller>
      )}
      {fieldType === 'Phone' && (
        <>
          <Controller
            control={control}
            name={toRegister}
            rules={{
              required: 'Phone number field cannot be empty.',
              minLength: {
                value: 7,
                message: 'Phone number must be between 7-15 characters.',
              },
              maxLength: {
                value: 15,
                message: 'Phone number must be between 7-15 characters.',
              },
            }}
            render={({
              field: { value, name, ref, onChange },
              fieldState: { error },
            }) => (
              <Form.Control
                type="number"
                style={{ fontSize: 13, height: 33 }}
                placeholder="Type here..."
                className={`${error ? 'is-invalid border-red' : ''}`}
                name={name}
                ref={ref}
                value={value}
                onChange={onChange}
              ></Form.Control>
            )}
          ></Controller>
        </>
      )}
```

**Reply by nimeshmaharjan1** [reply]
is there a way to access deps through the controller
